### PR TITLE
8318619: GenShen: Add configurable threshold for young triggers to expedite old evacuations

### PR DESCRIPTION
**Pull request by earthling-amzn**
When Shenandoah expects to free up space by promotions into young, or evacuations in old it may trigger a young collection. This change sets configurable thresholds for these conditions.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8318619](https://bugs.openjdk.org/browse/JDK-8318619): GenShen: Add configurable threshold for young triggers to expedite old evacuations (**Enhancement** - P4)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/347/head:pull/347` \
`$ git checkout pull/347`

Update a local copy of the PR: \
`$ git checkout pull/347` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/347/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 347`

View PR using the GUI difftool: \
`$ git pr show -t 347`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/347.diff">https://git.openjdk.org/shenandoah/pull/347.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/347#issuecomment-1773360172)

### CODE REVIEW

**Review by kdnilsen — APPROVED** [review]
Thanks.

File: src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp [hunk 1]
```diff
@@ -144,6 +144,16 @@
           " compact - run GC more frequently and with deeper targets to "   \
           "free up more memory.")                                           \
                                                                             \
+  product(uintx, ShenandoahExpeditePromotionsThreshold, 5, EXPERIMENTAL,    \
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp:147]: It's a good idea to add a Percent suffix to this.

File: src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp [hunk 2]
```diff
@@ -144,6 +144,16 @@
...
+          "When Shenandoah expects to promote at least this percentage "    \
+          "of the young generation, trigger a young collection to "         \
+          "expedite these promotions.")                                     \
+          range(0,100)                                                      \
+                                                                            \
+  product(uintx, ShenandoahExpediteMixedThreshold, 10, EXPERIMENTAL,        \
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoah_globals.hpp:153]: Similar to above, may be a good idea to add a Regions suffix to this.

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahYoungHeuristics.cpp [hunk 3]
```diff
@@ -150,8 +149,11 @@ bool ShenandoahYoungHeuristics::should_start_gc() {
 
   ShenandoahOldHeuristics* old_heuristics = heap->old_heuristics();
   size_t mixed_candidates = old_heuristics->unprocessed_old_collection_candidates();
-  if (mixed_candidates > 0) {
+  if (mixed_candidates > ShenandoahExpediteMixedThreshold && !heap->is_concurrent_weak_root_in_progress()) {
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/heuristics/shenandoahYoungHeuristics.cpp:152]: Might be a good idea to move the counting & test of threshold inside the test for collection state.

**Review by ysramakrishna — COMMENTED** [review]
LGTM, but left some pretty minor comments.

**Pull request closed without merging**